File: php/disk_usage.php
<?php

// cPanel API Configuration
$cpanel_host = "https://your-cpanel-domain:2087";
$username = "your_whm_username";
$api_token = "your_api_token";

// Function to interact with WHM API
function cpanel_api_request($endpoint, $params = [])
{
    global $cpanel_host, $username, $api_token;
    $url = "$cpanel_host/json-api/$endpoint?api.version=1";

    if (!empty($params)) {
        $url .= "&" . http_build_query($params);
    }

    $headers = [
        "Authorization: whm $username:$api_token",  
        "Accept: application/json"
    ];

    $ch = curl_init($url);
    curl_setopt($ch, CURLOPT_HTTPHEADER, $headers);
    curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
    curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, false); 
    curl_setopt($ch, CURLOPT_SSL_VERIFYHOST, false);

    $response = curl_exec($ch);
    if (curl_errno($ch)) {
        die("cURL Error: " . curl_error($ch));  
    }
    curl_close($ch);

    return json_decode($response, true); 
}

// Fetch all accounts
function get_cpanel_accounts()
{
    $response = cpanel_api_request("listaccts");

    if (!isset($response["data"]["acct"])) {
        die("Failed to fetch account list! Check API credentials.");
    }

    return $response["data"]["acct"];
}

// Process data
$accounts = get_cpanel_accounts();
$table_data = [];
$total_quota = 0;
$total_used = 0;

foreach ($accounts as $account) {
    $username = $account["user"];
    $domain = $account["domain"];
    $package = isset($account["plan"]) ? $account["plan"] : "N/A";  
    $quota = isset($account["disklimit"]) ? intval(str_replace("M", "", $account["disklimit"])) : "N/A";
    $used = isset($account["diskused"]) ? intval(str_replace("M", "", $account["diskused"])) : "N/A";

    if (is_numeric($quota) && is_numeric($used)) {
        $percent_used = $quota > 0 ? round(($used / $quota) * 100, 2) . "%" : "N/A";
        $total_quota += $quota;
        $total_used += $used;
    } else {
        $percent_used = "N/A";
    }

    $table_data[] = [$username, $domain, $package, $quota, $used, $percent_used];
}

// Total row (fixed position at the bottom)
$total_percent = $total_quota > 0 ? round(($total_used / $total_quota) * 100, 2) . "%" : "N/A";
$total_row = ["TOTAL", "-", "-", $total_quota, $total_used, $total_percent];

// ✅ Sort the table data by "Usage (%)" descending
usort($table_data, function ($a, $b) {
    $a_val = floatval(str_replace("%", "", $a[5]));
    $b_val = floatval(str_replace("%", "", $b[5]));
    return $b_val <=> $a_val;  // Sort descending
});

// ✅ Append TOTAL row at the end (always stays at the bottom)
$table_data[] = $total_row;

// ✅ Handle CSV download before any HTML output
if (isset($_GET['download']) && $_GET['download'] == 'csv') {
    header('Content-Type: text/csv');
    header('Content-Disposition: attachment; filename="cpanel_disk_usage.csv"');

    $output = fopen('php://output', 'w');
    fputcsv($output, ["Username", "Domain", "Package Used", "Quota (MB)", "Used (MB)", "Usage (%)"]);

    foreach ($table_data as $row) {
        fputcsv($output, $row);
    }

    fclose($output);
    exit;
}
?>

<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>cPanel Disk Usage</title>
    <style>
        body { font-family: Arial, sans-serif; text-align: center; background-color: #f4f4f9; padding: 20px; }
        h2 { color: #333; }

        /* Auto-size table width based on content */
        table { 
            margin: 20px auto; 
            border-collapse: collapse; 
            width: auto;
            background: #fff; 
            border-radius: 8px; 
            overflow: hidden; 
            box-shadow: 0px 2px 10px rgba(0, 0, 0, 0.1);
            max-width: 90%;
            min-width: 500px;
        }
        
        th, td { padding: 12px; text-align: left; border-bottom: 1px solid #ddd; }
        th { background-color: #007bff; color: white; cursor: pointer; }
        th:hover { background-color: #0056b3; }
        tr:nth-child(even) { background-color: #f9f9f9; }
        tr:hover { background-color: #f1f1f1; }
        .total-row { font-weight: bold; background-color: #007bff !important; color: white; }
        .download { margin-top: 20px; }
        button { border: none; background: #007bff; color: white; padding: 10px 15px; border-radius: 5px; cursor: pointer; }
        button:hover { background: #0056b3; }
    </style>
    <script>
        function downloadCSV() {
            window.location.href = "?download=csv";
        }
    </script>
</head>
<body>

<h2>cPanel Disk Usage Report</h2>

<table id="cpanelTable">
    <tr>
        <th onclick="sortTable(0)">Username</th>
        <th onclick="sortTable(1)">Domain</th>
        <th onclick="sortTable(2)">Package Used</th>
        <th onclick="sortTable(3)">Quota (MB)</th>
        <th onclick="sortTable(4)">Used (MB)</th>
        <th onclick="sortTable(5)">Usage (%)</th>
    </tr>
    <?php foreach ($table_data as $row): ?>
    <tr class="<?php echo ($row[0] === 'TOTAL') ? 'total-row' : ''; ?>">
        <td><?php echo htmlspecialchars($row[0]); ?></td>
        <td><?php echo htmlspecialchars($row[1]); ?></td>
        <td><?php echo htmlspecialchars($row[2]); ?></td>
        <td><?php echo htmlspecialchars($row[3]); ?></td>
        <td><?php echo htmlspecialchars($row[4]); ?></td>
        <td><?php echo htmlspecialchars($row[5]); ?></td>
    </tr>
    <?php endforeach; ?>
</table>

<div class="download">
    <button onclick="downloadCSV()">Download CSV</button>
</div>

</body>
</html>

File: python/disk_usage.py
import requests
import pandas as pd
import urllib3

# Suppress SSL warnings (optional, but not recommended)
urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)

# Configuration (Replace with your cPanel details)
CPANEL_HOST = "https://your-cpanel-domain:2087"  # Replace with your WHM/cPanel URL
USERNAME = "your_whm_username"
API_TOKEN = "your_api_token"

HEADERS = {
    "Authorization": f"whm {USERNAME}:{API_TOKEN}"
}

# Function to get account list
def get_cpanel_accounts():
    url = f"{CPANEL_HOST}/json-api/listaccts?api.version=1"
    response = requests.get(url, headers=HEADERS, verify=False)  
    data = response.json()
    
    if "data" in data and "acct" in data["data"]:
        return data["data"]["acct"]
    return []

# Function to get disk usage
def get_disk_usage(username):
    url = f"{CPANEL_HOST}/json-api/accountsummary?api.version=1&user={username}"
    response = requests.get(url, headers=HEADERS, verify=False)

    try:
        data = response.json()

        if "data" in data and "acct" in data["data"]:
            account_data = data["data"]["acct"][0]
            quota = account_data.get("disklimit", "N/A")  # Get quota
            used = account_data.get("diskused", "N/A")  # Get used space

            # Convert quota if not actually unlimited
            if quota.lower() == "unlimited":
                quota = "Unlimited"
                percent_used = "N/A"
            else:
                quota = int(quota.replace("M", ""))  # Convert '40960M' -> 40960
                used = int(used.replace("M", ""))  # Convert '37240M' -> 37240
                percent_used = round((used / quota) * 100, 2) if quota > 0 else "N/A"

            return quota, used, percent_used

    except Exception:
        pass  # Suppressing verbose errors

    return "N/A", "N/A", "N/A%"

# Fetch accounts and build table
accounts = get_cpanel_accounts()
table_data = []
total_quota = 0
total_used = 0

for account in accounts:
    username = account.get("user")
    domain = account.get("domain")
    quota, used, percent_used = get_disk_usage(username)

    # Sum up quota and used space (skip "Unlimited" cases)
    if isinstance(quota, int):
        total_quota += quota
    if isinstance(used, int):
        total_used += used

    table_data.append([username, domain, quota, used, f"{percent_used}%"])

# Calculate total percentage
total_percent = round((total_used / total_quota) * 100, 2) if total_quota > 0 else "N/A"

# Add total row
table_data.append(["TOTAL", "-", total_quota, total_used, f"{total_percent}%"])

# Create DataFrame and display
df = pd.DataFrame(table_data, columns=["Username", "Domain", "Quota (MB)", "Used (MB)", "Usage (%)"])
print(df)  # Print in console
df.to_csv("cpanel_accounts.csv", index=False)  # Save to CSV
print("Data saved to cpanel_accounts.csv")